Adds test that checks ReadingGoalService save method for null safety.

dmitraver opened the pull request:
## Summary of change

Checks for null safety when using ReadingGoalService save method.

## Related issue

Closes #375 

## Pull request checklist

Please keep this checklist in & ensure you have done the following:

- [x] Read, understood and adhered to our [contributing document](https://github.com/knjk04/book-project/blob/master/CONTRIBUTING.md).
  - [x] Ensure that you were first assigned to a relevant issue before creating this pull request
  - [x] Ensure code changes pass all tests

- [x] Read, understood and adhered to our [style guide](https://github.com/knjk04/book-project/blob/master/STYLEGUIDE.md). A lot of our code reviews are spent on ensuring compliance with our style guide, so it would save a lot of time if this was adhered to from the outset. 

- [x] Filled in the summary, context (if applicable) and related issue section. Replace the square brackets and its placeholder content with your contents. For an example, see any merged in pull request
  - [x] Included a screenshot(s) if a UI change was involved (it may look different on a reviewer's device)

- [x] Created a branch that has a descriptive name (what your branch is for in a few words and includes the issue number at the end, e.g. `test-reading-goal-123`

- [x] Set this pull request to 'draft' if you are still working on it

- [x] Resolved any merge conflicts

If in doubt, get in touch with us via our [Slack workspace](https://join.slack.com/t/teambookproject/shared_invite/zt-h6vx3n8l-KN8QnO50r7QWZHgHFnFdPw)

devsnice commented:
@dmitraver @knjk04  why do you write such pointless tests and checks? The good design is not expect null as parameters at all, for it application should be designed with protection from this situation. How null can be passed to ReadingGoalService and why? Where is the places when we can get null before it? If there's no such record in database you can throw NotFoundException on place and not go further to ReadingGoalService with nullish values.

dmitraver commented:
@devsnice "The good design is not expect null as parameters at all". You can expect anything but the reality is that someone would eventually pass null value and we want usually to fail fast and not let it go through all the layers up to the database and fail somewhere there. Don't rely on users for giving you always properly validated data. I'm strongly against ignoring validation and let it fail in some obscure places down the road.

knjk04 commented:
@devsnice 

> why do you write such pointless tests and checks?

I appreciate this may not have been your intention, but this is inappropriate. This is a violation of our [code of conduct](https://github.com/Project-Books/book-project/blob/master/CODE_OF_CONDUCT.md). The issue here is that you're targeting specific people with language that is not welcoming. I'd like to remind you that in the code of conduct, we specifically mention that we strive to create a welcoming environment regardless of "level of experience' (#375 is labelled as 'good first issue').

If you would like to follow up on this, the best place to do so would be to message me privately on Slack, not here. 

> The good design is not expect null as parameters at all, for it application should be designed with protection from this situation.

Generally speaking, I agree with this. 

> How null can be passed to ReadingGoalService and why?

It can be passed to the `save()` method, but I agree that it shouldn't (I believe we check for this). Perhaps you're right in that there is a better way of going about this. If so, we'd be happy to find out what your suggestion is. My initial thinking was that since we don't want nulls in the database, we should check that we're not doing accidentally doing this. By implementing this test, we are notified, by virtue of a failing test, if such a thing is ever made permissible (which, as it stands, it is).

> If there's no such record in database you can throw NotFoundException on place and not go further to ReadingGoalService with nullish values.

I also agree with this. However, this issue is dealing with a save operation, not a find operation. We cannot throw such an exception ourselves given that we're overriding the method. Not found exceptions are thrown in scenarios when searching for something that doesn't exist (this is handled for us).

> Another option is consider Optional class and check value (Optional::map), before the method invocation.

We're using `Optional` in other places (and are extending our use of it). I don't think we can pass an `Optional` to the `save()` method because by doing so, it would not override the superclass.

Code quality is important to us, so we appreciate your suggestions and comments. Feel free to follow up on any of the code replies I have given. It may be the case that I have misunderstood you. 

A part of being open to writing better quality code is understanding that we're constantly learning and improving. We don't have the highest quality code, but we strive to this goal (which, arguably, is unachievable) through constant refactoring. I'd encourage you to make future suggestions because it's really important, but please feel free to respectfully critique the code, not the person/people writing it.

knjk04 commented:
@devsnice @dmitraver 

Since this is both an open-ended and controversial discussion (some argue against using `Optional`s for parameters), I'd say let's not continue the discussion here (@dmitraver's reply only just came through for me). We'll go ahead with this PR. 

The best place to continue this discussion (while adhering to both the spirit and letter of our code of conduct) would be over on Slack.

knjk04 reviewed (changes requested):
Thanks for working on this! This works as intended. 

I've made a suggestion on renaming the method for clarity.

devsnice commented:
> @devsnice "The good design is not expect null as parameters at all". You can expect anything but the reality is that someone would eventually pass null value and we want usually to fail fast and not let it go through all the layers up to the database and fail somewhere there

the reality depends only on how team maintains code, if it starts from not going this way, the way is going to be without any null checking. Project can have recommendation to not return nulls anywhere 
1) methods should not return null
2) variables should not initialized with null

it's basically main rules to follow to have strong codebase integrity that allows us to not check any code on null.

> Don't rely on users for giving you always properly validated data. I'm strongly against ignoring validation and let it fail in some obscure places down the road.

in this case we need to validate data not inside methods, but in place where we can get null, basically in http-resources, where we can get some unpredicatable request-params

devsnice commented:
@knjk04 thanks for your message, I wasn't intended to write in disrespectful tone, sorry if it looks like it, I will edit it.
>  I also agree with this. However, this issue is dealing with a save operation, not a find operation. We cannot throw such an exception ourselves given that we're overriding the method. Not found exceptions are thrown in scenarios when searching for something that doesn't exist (this is handled for us).

we can check it inside http-request, if user send valid data, we're able to create valid object and we will able to save it to db

knjk04 commented:
@devsnice Thanks for your comments & apology.

> Since this is both an open-ended and controversial discussion (some argue against using Optionals for parameters), I'd say let's not continue the discussion here (@dmitraver's reply only just came through for me). We'll go ahead with this PR.
The best place to continue this discussion (while adhering to both the spirit and letter of our code of conduct) would be over on Slack.

In line with my previous comment, let's leave this here for now. This open-ended discussion is better over Slack. If we're happy to proceed with your recommendations, we can create a new issue to resolve this.

knjk04 commented:
@dmitraver Let us know when you've addressed the review feedback. Also, could you please the resolve the merge conflict?

dmitraver commented:
Sure thing, I will make changes soon.

On Sun, 4 Oct 2020 at 20:12, Karan Kumar <notifications@github.com> wrote:

> @dmitraver <https://github.com/dmitraver> Let us know when you've
> addressed the review feedback. Also, could you please the resolve the merge
> conflict?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/Project-Books/book-project/pull/379#issuecomment-703293850>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AANXADEMUFIKSQOZBCEAU5TSJC3IPANCNFSM4SAEASAA>
> .
>


-- 
Best wishes,
Dmitry Avershin.

dmitraver commented:
@knjk04 fixed the name, conflicts and squashed commits. I had to adjust test as well. Please take a look again and let me know if I need to change anything there.

knjk04 reviewed (approved):
Thanks for working on this, Dmitry!
